packages/@expo/cli/src/start/server/metro/inspector-proxy/device.ts

```diff
@@ -0,0 +1,50 @@
+import type { DebuggerInfo, Device as MetroDevice } from 'metro-inspector-proxy';
+import type WS from 'ws';
+
+import {
+  NetworkResponseHandler,
+  NetworkReceivedResponseBody,
+  NetworkGetResponseBody,
+} from './messages/NetworkRespose';
+import { DeviceRequest, DebuggerRequest } from './messages/types';
+
+type KnownDeviceRequests = DeviceRequest<NetworkReceivedResponseBody>;
+type KnownDebuggerRequests = DebuggerRequest<NetworkGetResponseBody>;
+
+export function createInspectorDeviceClass(MetroDeviceClass: typeof MetroDevice) {
+  return class ExpoInspectorDevice extends MetroDeviceClass {
+    networkResponseBodyHandler = new NetworkResponseHandler();
+
+    onDeviceMessage(request: KnownDeviceRequests, debuggerInfo: DebuggerInfo): void {
+      switch (request.method) {
+        case 'Expo(Network.receivedResponseBody)':
+          return this.networkResponseBodyHandler.onDeviceMessage(request);
+      }
+    }
+
+    onDebuggerMessage(request: KnownDebuggerRequests, debuggerInfo: DebuggerInfo, socket: WS): any {
+      switch (request.method) {
+        case 'Network.getResponseBody':
+          return this.networkResponseBodyHandler.onDebuggerMessage(request);
+      }
+    }
+
+    /** Hook into the message life cycle to answer more complex CDP messages */
+    _processMessageFromDevice(message: DeviceRequest<any>, debuggerInfo: DebuggerInfo) {
+      this.onDeviceMessage(message, debuggerInfo);
+      return super._processMessageFromDevice(message, debuggerInfo);
+    }
+
+    /** Hook into the message life cycle to answer more complex CDP messages */
+    _interceptMessageFromDebugger(
+      request: DebuggerRequest<any>,
+      debuggerInfo: DebuggerInfo,
+      socket: WS
+    ) {
+      const result = this.onDebuggerMessage(request, debuggerInfo, socket);
+      return result
```

Comment:
i see a problem here that metro's `_interceptMessageFromDebugger` returns `boolean` only: https://github.com/facebook/metro/blob/4c520eda2ef322b1b8057d880a13ed776ecd3e79/packages/metro-inspector-proxy/src/Device.js#L457 and it doesn't send out the command returned in `result`. so from the devtools' protocol monitor, the `Network.getResponseBody` will not receive response through websocket.

Reply:
When adding vscode compatibility, I noticed this too. Will revise the types and docs to reflect better what methods should return! Thanks for flagging this

Reply:
Fixed, I also made the types a bit stricter around `Device`, cdp messages, and used `devtools-protocol` types in the network handler. Should get us better TS warnings related to protocol changes too :)